Add a build workflow

keithchev opened the pull request:
This PR adds a Github actions workflow (`build.yml`) to "build" the publication by aggregating the necessary files from the git commit history.

It also slightly modifies the existing `publish.yml` workflow to make it more robust by creating the required `gh-pages` branch if it does not exist.

The logic behind the `build.yml` workflow is a little complex. I attempted to document it and also the overall publication workflow in the `GITHUB_ACTIONS.md` file added with this PR. FWIW, I would recommend to start reviewing this PR by reading that document first.

An example of the PRs that are opened by the `build.yml` workflow can be seen [here](https://github.com/Arcadia-Science/2025-notebook-pub-dev/pull/12).

## Testing

These workflows are hard to test, as they manipulate branches, open PRs, are triggered by PRs, and depend on the existence of git tags. I've tested them manually in a separate temporary repo (created from this template repo), and I *think* they're working correctly. But I also think we should test them again, along with the entire publication workflow, once we merge this PR.

ekiefl commented on developer-docs/GITHUB_ACTIONS.md:
```diff
@@ -0,0 +1,52 @@
+
+## The version aggregation problem
+
+The challenge addressed by the `build.yml` workflow is **version aggregation**. This refers to the fact that the public Quarto site (that is, the public publication) needs to show multiple historical versions of the notebook `index.ipynb`, one for each tagged release in the GitHub repository. Because GitHub tags correspond to commits, these versions exist in different commits in the repository. T ...
```
This document is a great paper trail. Nicely done

ekiefl commented on developer-docs/GITHUB_ACTIONS.md:
```diff
@@ -0,0 +1,52 @@
+
+1. It runs the [`_build.py`](../_build.py) script. This script:
+   - Extracts the `index.ipynb` notebook from each tag in the repository.
+   - Copies each version to a new uniquely named file (e.g., `index_v1.ipynb`, `index_v2.ipynb`).
```
Might need to update this since the latest version we will rename to `index.ipynb`

ekiefl reviewed (commented):
Amazing job on the implementation and documentation. Thank you! Please re-request once you make the file rename changes.

keithchev commented on _build.py:
```diff
@@ -0,0 +1,126 @@
+        return
+
+    if os.path.exists(dst):
+        shutil.rmtree(dst)
```
I copied this from `bump_version` 😉. I don't think `rmtree` works on files; I've fixed this and also switched to pathlib where possible.

keithchev commented:
@ekiefl I've made the changes we discussed around `index.ipynb`, addressed your comments, and tested the build script "for real" in my toy repo (and caught a bug 😬).

ekiefl reviewed (approved):
This is such a game-changer for the versioning capabilities of these pubs. It will be great to have git tags glue together both Zenodo archives and the current pub version.

wasimsandhu commented on developer-docs/GITHUB_ACTIONS.md:
```diff
@@ -0,0 +1,53 @@
+
+The challenge addressed by the `build.yml` workflow is **version aggregation**. This refers to the fact that the public Quarto site (that is, the public publication) needs to show multiple historical versions of the notebook `index.ipynb`, one for each tagged release in the GitHub repository. Because GitHub tags correspond to commits, these versions exist in different commits in the repository. T ...
+
+## The build workflow (`build.yml`)
```
This is a major nit, but I think calling this a build workflow is somewhat misleading, since it's actually just copying and renaming files/directories, and not actually building the Quarto site.

I think "version aggregation workflow" (or something similar) would be a more accurate title for this workflow.

keithchev replied:
I dunno. I see what you mean but I wanted a simple name that's equivalent to the "publish" name for the Quarto workflow, and fwiw it does more or less feel like a build step to me, because it is creating the files required to render the Quarto site. This feels analogous to a compilation or transpilation step (like calling `tsc` to transpile typescript). The fact that "creating the files" happens to be trivial in this case (because it's mostly just copying/renaming files) is an implementation detail, imo.

keithchev replied:
Following up: we chatted offline and agreed the "build" name is not ideal but agreed not to rename this for now.

wasimsandhu commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,80 @@
+
+      # Reset the `publish` branch to point to `main`, because we use the `publish` branch
+      # only to store the build artifacts, not for version control.
+      - name: Reset the publish branch
```
Nit: In case an error occurs in the `_build.yml` step, maybe this step should be performed after that script is run, and before the PR is opened?

keithchev replied:
Yeah, I hear you, but I think this step has to come before the `_build.py` script is called because that script modifies the working tree, which may make it impossible to later check out the publishing branch in this step. That's why all the git-related steps come first, and the `_build.py` script is called right before the final step to open a PR.

wasimsandhu reviewed (approved):
These workflows are really great to have. Major props for setting this up and testing the entire workflow yourself, presumably many times!

All of my comments are nits or questions, so I'm approving this PR as-is. LGTM.

wasimsandhu commented on _build.py:
```diff
@@ -0,0 +1,158 @@
+
+    try:
+        subprocess.run(["git", "checkout", ref], check=True)
+        yield
```
Update: @ekiefl just explained to me how this works and why it's necessary. Nevermind! 🙂

keithchev commented on _build.py:
```diff
@@ -0,0 +1,158 @@
+    freeze_src.replace(freeze_dst)
+
+
+def update_quarto_yaml(most_recent_tag: str, previous_tags: list[str], dry_run: bool) -> None:
```
It's just a way to run the script without actually doing anything. I added this to the docstring.

keithchev commented on developer-docs/PUBLISHING_GUIDE.md:
````diff
@@ -72,28 +82,44 @@ Instructions are provided for:
+Then, create a new tag and push it to the repo:
+
+```bash
+git tag v2
````
yep, good call.
